### PR TITLE
Permissions rework

### PR DESCRIPTION
**Pull request by tgardiner**

### CODE REVIEW

File: cloudwatch_logs/lambda/cloudwatch_subscriber/main.py [hunk 1]
```diff
@@ -15,31 +15,39 @@
...
 CWLOGS = boto3.client('logs')
+# Create an AWS Lambda client, to interact with the Lambda API
+LAMBDA = boto3.client('lambda')
 
 # Sends a FAILED response to CloudFormation X seconds before the lambda times out
 LAMBDA_TIMEOUT = 15
 
+# A seed for later UUID calls as we need a reproducible way to produce the StatementId
+# for Lambda the permission policy
+UUID_SEED = UUID('e4ce7ea4-1343-427a-9e51-a22aad3dd0a8')
```
Comment (czerwingithub) [inline cloudwatch_logs/lambda/cloudwatch_subscriber/main.py]: Won't a fixed seed mean that we are generating the same UUIDs across Lambda invocations?

File: cloudwatch_logs/lambda/cloudwatch_subscriber/main.py [hunk 2]
```diff
@@ -198,6 +206,62 @@ def diff_log_groups(log_group_options, old_log_group_options):
...
+            Action='lambda:InvokeFunction',
+            Principal=f"logs.{region}.amazonaws.com",
+            SourceArn=f"arn:aws:logs:{region}:{account_id}:log-group:{log_group_name}:*",
+            SourceAccount=account_id
+        )
+    except LAMBDA.exceptions.ResourceConflictException as e:
+        LOGGER.info(f"Error adding lambda permission: {log_group_name}: {e}")
+    except:
+        LOGGER.exception(f"Error adding lambda permission: {log_group_name}")
+        raise
```
Comment (czerwingithub) [inline cloudwatch_logs/lambda/cloudwatch_subscriber/main.py]: Why do you raise the exception here but not in the catch for `ResourceConflictException` a few line aboves?

File: cloudwatch_logs/lambda/cloudwatch_subscriber/main.py (hunk 1 above)
Comment (tgardiner) [reply cloudwatch_logs/lambda/cloudwatch_subscriber/main.py]: Yeah that the goal, we need a reproducible way of generating an alphanumeric string to be used as the StatementId in the lambda_add_permission and lambda_remove_permission functions.  Both the add_permission and remove_permission calls to the AWS API need to use the same StatementId to add/remove the permission statement.  

I've leveraged the UUID library here as it produces a hex string from the result which I can use for the StatementId without further modification.  It's essentially just SHA1-hashing the provided seed together with the logGroup name to produce the result.

File: cloudwatch_logs/lambda/cloudwatch_subscriber/main.py (hunk 2 above)
Comment (tgardiner) [reply cloudwatch_logs/lambda/cloudwatch_subscriber/main.py]: A resource conflict exception here is the result of the StatementId already existing, which means we must have already added the permission on a previous run.  I've added a comment to make that a little more obvious.  The exact error is:

```
An error occurred (ResourceConflictException) when calling the AddPermission operation: The statement id (CloudWatchLogs) provided already exists. Please provide a new statement id, or remove the existing statement.
```

Comment (czerwingithub) [reply cloudwatch_logs/lambda/cloudwatch_subscriber/main.py]: Ok, I would then change the error message to something like `Warning, lambda permission already exists.  No action taken.`  Essentially, something to say that this call didn't do anything but telling the user that's probably ok.

Comment (tgardiner) [reply cloudwatch_logs/lambda/cloudwatch_subscriber/main.py]: Understood, I've updated that and the other non-critical messages to match

**Pull request merged**